sway-lsp/src/core/token_map.rs

```diff
@@ -57,30 +57,72 @@ impl TokenMap {
 
     /// Given a cursor [Position], return the [Ident] of a token in the
     /// Iterator if one exists at that position.
-    pub fn ident_at_position<I>(&self, cursor_position: Position, tokens: I) -> Option<Ident>
+    pub fn idents_at_position<I>(&self, cursor_position: Position, tokens: I) -> Vec<Ident>
     where
         I: Iterator<Item = (Ident, Token)>,
     {
-        for (ident, _) in tokens {
-            let range = token::get_range_from_span(&ident.span());
-            if cursor_position >= range.start && cursor_position <= range.end {
-                return Some(ident);
-            }
-        }
-        None
+        tokens
+            .filter_map(|(ident, _)| {
+                let range = token::get_range_from_span(&ident.span());
+                if cursor_position >= range.start && cursor_position <= range.end {
+                    return Some(ident);
+                }
+                None
+            })
+            .collect()
     }
 
-    /// Check if the code editor's cursor is currently over one of our collected tokens.
+    /// Returns the first collected tokens that is at the cursor position.
     pub fn token_at_position(&self, uri: &Url, position: Position) -> Option<(Ident, Token)> {
         let tokens = self.tokens_for_file(uri);
-        self.ident_at_position(position, tokens).and_then(|ident| {
-            self.try_get(&token::to_ident_key(&ident))
-                .try_unwrap()
-                .map(|item| {
-                    let ((ident, _), token) = item.pair();
-                    (ident.clone(), token.clone())
-                })
-        })
+        self.idents_at_position(position, tokens)
+            .first()
+            .and_then(|ident| {
+                self.try_get(&token::to_ident_key(ident))
+                    .try_unwrap()
+                    .map(|item| {
+                        let ((ident, _), token) = item.pair();
+                        (ident.clone(), token.clone())
+                    })
+            })
+    }
+
+    /// Returns all collected tokens that are at the cursor position.
+    pub fn tokens_at_position(
```

Comment:
I'm not sure I fully understand the intended context of this change. There should only ever be a single token at a given position. Can you explain a bit more about what your intention is here?

Reply:
Sometimes there are multiple idents at a position. For instance, with the expression `something.`, where the cursor position is before the `.`, the tokens returned are `something` and `.`

This is because we're doing an inclusive comparison of the tokens' ranges. Previously `ident_at_position` would just return the first ident found, discarding the rest. I changed it to `idents_at_position` to return all of the idents found, and updated the comment of `token_at_position` to reflect this.

The reason for the `tokens_at_position` function is that we want to find tokens that have this position in the _body_ of the token, not necessarily the ident of the token. If we want to get the token for the function declaration that the cursor position is inside of, searching with the span of the ident of the function declaration won't be enough. 

For example:

```
fn fun() {
  let something = MyStruct::new();
  something. // <-- cursor position is before the `.`
}
```

I'm using `tokens_at_position` to find the declaration of `fun` so that I can look inside its body and parameters to see what type `something` is. If I used `idents_at_position` I would only get back the ErrorRecovery tokens (`something` and `.`), but I need to know the contextual token that contains it (the function declaration).

Reply:
Ok that makes sense. Maybe it's worth elaborating on the doc comment the intended use case just so it's a bit clearer.